File: 卡码/代码随想录/JS/图论/bfs/LeetCode_0127.js
//127. 单词接龙
const ladderLength = function(beginWord, endWord, wordList) {
  const visited=new Set();
  const wordSet=new Set(wordList);
  if(!wordSet.has(endWord)) return 0;
  const q=new Array();
  let path=1;
  q.push(beginWord);
  const changeLetter=(word)=>{
    for(let i=0;i<word.length;i++){
      for(let ch="a".charCodeAt(0);ch<="z".charCodeAt(0);ch++){
        const newWord=word.slice(0,i)+String.fromCharCode(ch)+word.slice(i+1);
        if(endWord==newWord) return true;
        if(wordSet.has(newWord)&&!visited.has(newWord)){
          visited.add(newWord);
          q.push(newWord);
        }
      }
    }
    return false;
  }
  while(q.length>0){
    const size=q.length;
    for(let i=0;i<size;i++){
      const word=q.shift();
      if(changeLetter(word)) return path+1;
    }
    path++;
  }
  return 0;
};

File: 卡码/代码随想录/1/11-图论/Java/src/com/kurumi/bfs/LeetCode_0127.java
//127. 单词接龙
package com.kurumi.bfs;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class LeetCode_0127 {

}
class Solution {
	HashSet<String> wordSet;
	HashSet<String> visited=new HashSet();
	Deque<String> q=new ArrayDeque<>();
	public boolean changeLetter(String word,String target) {
		char[] chars = word.toCharArray();
		for(int i=0;i<chars.length;i++) {
			char origin=chars[i];
			for(char ch='a';ch<='z';ch++) {
				chars[i]=ch;
				String newWord = String.valueOf(chars);
				if(target.equals(newWord)) return true;
				if(wordSet.contains(newWord)&&!visited.contains(newWord)) {	//单词集合包含新单词且未被访问过
					visited.add(newWord);
					q.offer(newWord);
				}
			}
			chars[i]=origin;	//回溯
		}
		return false;
	}
  public int ladderLength(String beginWord, String endWord, List<String> wordList) {
  	wordSet=new HashSet<>(wordList);
  	if(!wordSet.contains(endWord)) return 0;	//未找到终点字符串
  	visited.add(beginWord);
  	q.offer(beginWord);
  	int path=1;
  	while(!q.isEmpty()) {
  		int size=q.size();
  		for(int i=0;i<size;i++) {
  			String word = q.poll();
  			if(changeLetter(word,endWord)) return path+1;
  		}
  		path++;
  	}
  	return 0;
  }
}